### ISSUE TITLE
Atypical long literal results in error

### ISSUE THREAD

**Issue by LaurensRietveld** [rdfjs/N3.js#128]
Error you get is the following:

```
RangeError: Maximum call stack size exceeded
    at RegExp.exec (<anonymous>)
    at N3Lexer._tokenizeToEnd (/home/lrd900/triply/N3.js/lib/N3Lexer.js:161:51)
    at ReadStream.<anonymous> (/home/lrd900/triply/N3.js/lib/N3Lexer.js:405:16)
    at emitOne (events.js:116:13)
    at ReadStream.emit (events.js:211:7)
    at addChunk (_stream_readable.js:263:12)
    at readableAddChunk (_stream_readable.js:246:13)
    at ReadStream.Readable.push (_stream_readable.js:208:10)
    at fs.read (fs.js:2042:12)
    at FSReqWrap.wrapper [as oncomplete] (fs.js:658:17)
```

To reproduce: `node ./perf/N3Lexer-perf.js ./out_1000000_points.trig`
See attached for a zipped [example file](https://github.com/RubenVerborgh/N3.js/files/1775876/out_1000000_points.zip)

**Comment by RubenVerborgh** [comment]
Thanks for reporting.

So the native `RegExp.exec` is causing this? Then there might not be much I can do about this I'm afraid… Maybe a slightly different regex fixes this (that would be good); the alternative would be to do this without regexes, but that would totally kill performance.

**Comment by wouterbeek** [comment]
The lexical form of this literal contains 28,802,110 characters.  I have experience with storing lexical forms of up to ~1M characters using N3, so the limit occurs somewhere between 1M and 29M.

Unfortunately, lengthy literals have real word applications, e.g., when encoding a detailed polygon of a large country of irregular shape.

**Comment by RubenVerborgh** [comment]
> 28,802,110 characters

Oh. That seems quite extreme from the perspective of V8.
Seems like we're walking straight into https://bugs.chromium.org/p/v8/issues/detail?id=3878

We should see if it's possible to use a different regex that avoids backtracking (background: https://www.regular-expressions.info/catastrophic.html).

**Comment by RubenVerborgh** [comment]
In general, the problem is that we are hitting V8 limits.

However, in this specific case, the literal is actually simple enough to pass the more simple regular expression, which is tested first. Yet when parsing this file in streaming mode, the more simple regex will fail if the file is only read partially, so it falls through to the more complex regex, which crashes.

f529ff9197a65aaa215f4968e0b2f71dc6ba3e0a adds a guard so we only fall through to the complex cases if there is a chance of them matching.

Can you verify this solution works for you, and reopen if it doesn't?

**Comment by LaurensRietveld** [comment]
@RubenVerborgh I just hit this limit again, for the attached file. This seems to be a case where the simple regex does not match, and where we're hitting the V8 limits for the complex regex. 
Not expecting a solution here, but posting it for documentation purposes
[long.nt.gz](https://github.com/rdfjs/N3.js/files/3863768/long.nt.gz)

**Comment by RubenVerborgh** [comment]
Thanks for this!
I have to backtrack on my earlier statement that

> Then there might not be much I can do about this I'm afraid…

Not with regexes for sure, but we could fallback to a char-by-char, non-regex solution if we detect that the literal is getting too large.

**Comment by RubenVerborgh** [comment]
@LaurensRietveld Traced it back to a single line; it's an 8MB literal 😂
But sure, if valid RDF, we will parse it.

**Comment by LaurensRietveld** [comment]
Yes, this seems to be the underappreciated contribution of John Galsworthy to Linked Data ;)

**Comment by RubenVerborgh** [comment]
Published [v1.3.1 AKA the John Galsworthy release](https://github.com/rdfjs/N3.js/releases/tag/v1.3.1). Thanks a lot, @LaurensRietveld and @WouterBeek.

**Comment by RubenVerborgh** [comment]
@LaurensRietveld The case for `long.nt` was fixed, and I also added some performance improvements. It was a very tricky case where the literal contained millions of characters and 50.000 quotation marks. We should be resilient against _any_ literal now, as long as it fits in main memory (let me know when you hit _that_ limit).

However, I wondered whether you have any control over the serialization of files like `long.nt`, or whether you have found them on the wild. Because if you do have any control, I strongly recommend serializing literals containing quotes using the `"""…"""` syntax if Turtle is permissible. Because in single-quote strings, _any_ RDF parser (including N3.js, but all others) have to examine every single quote, whereas in triple-quoted strings, only sequences of 3 quotes need to be examined.

This also happens because N3.js does not have a dedicated N-Triples mode, where one can make stronger assumptions about lines.
